### ISSUE TITLE
New rule proposal: namespace must match file location

### ISSUE THREAD

**Issue by oatkins** [DotNetAnalyzers/StyleCopAnalyzers#1481]
It's common to follow the VS standard of have all files contain a single namespace that corresponds to the default namespace for the project, followed by the names of the sub-directories that contain the file. It's easy to forget to fix the namespace after moving a file. This proposal is to add an analyzer that reports the incorrect namespace.

Sadly this may not be easy, since I don't know that Roslyn grants access to the "default namespace" property of a project.

**Comment by Maxwe11** [comment]
:+1:

**Comment by vweijsters** [comment]
I'm undecided on this one. I know quite a few people who would love it, but personally I don't like it all. As a developer I should decide the namespaces for types, without being bound to a specific layout that is due to the way that I organize my source code. This is especially important when creating assemblies with a clearly defined public API.

I would support such a rule only if it is disabled by default.

**Comment by vweijsters** [comment]
It may be possible to retrieve the default namespace using the `IGenerateTypeService.GetRootNamespace` method.
```c#
var generateTypeService = document.GetLanguageService<IGenerateTypeService>();
var rootNamespace = generateTypeService.GetRootNamespace(document.Project.CompilationOptions);
```

Note that this is purely theoretical, I've haven't tested anything.

**Comment by pdelvo** [comment]
I dont think that this will work. the C# implementation for GetRootNamespace returns an empty string http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/GenerateType/CSharpGenerateTypeService.cs,9d908aef53327616,references

**Comment by vweijsters** [comment]
Agree, that's why I deleted the comment. Another problem is that the stuff is all internal, and therefore not accessible.

**Comment by oatkins** [comment]
I had a quick go at writing this in my own project, but I bumped up against the same restrictions. Any suggestions would be welcome!

**Comment by sharwell** [comment]
Well, we could always **stylecop.json**. :worried:

**Comment by oatkins** [comment]
@sharwell That would mean having to use a different **stylecop.json** in each project in a solution. Currently it's convenient to use linked files to share one settings file between all projects. (This is especially relevant because we don't have the settings file merging feature from legacy StyleCop.)

**Comment by sharwell** [comment]
Someone should file a feature request for Roslyn and link it back to this issue.

**Comment by oatkins** [comment]
Here you go: https://github.com/dotnet/roslyn/issues/5252

**Comment by sharwell** [comment]
This is now being tracked in dotnet/roslyn#3798.

**Comment by mwelsh1118** [comment]
The Roslyn issue is now closed. Is this StyleCop rule in the works?

**Comment by MrMikeJJ** [comment]
This rule is now implemented in Visual Studio https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0130

**Comment by PTwr** [comment]
There is one undocumented limitation to IDE0130, it is suppressed if namespace contains partial classes. Even if all parts of class are in same directory. (But its smart enough to still fire on not-really-partial classes)
Which makes it half-useless when you try to cleanup namespaces in WinForms project. 
There appears to be no option to change that behavior (guess how I found this issue...) so having more configurable one in SA would be nice to have.
